extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1339,57 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
-    ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (this.isPaywallStory_()) {
+      if (
+        pageIndex >= PAYWALL_PAGE_INDEX &&
+        subscriptionsState !== SubscriptionsState.GRANTED
+      ) {
+        // Block while waiting for upcoming entitlement.
+        if (subscriptionsState === SubscriptionsState.UNKNOWN) {
+          return this.subscriptionsStatePromise_.promise.then(() => {
+            return this.switchTo_(targetPageId, direction);
+          });
+        }
+        if (!this.pageAfterGranted_) {
+          this.pageAfterGranted_ = targetPageId;
+        }
+
+        // Attempt to navigate to the locked pages after the paywall page(the first blocked page) with blocked state.
+        if (pageIndex > PAYWALL_PAGE_INDEX) {
+          // Show the paywall immediately if the active page is the paywall page.
+          if (this.getPageIndex(this.activePage_) === PAYWALL_PAGE_INDEX) {
+            this.storeService_.dispatch(
+              Action.TOGGLE_SUBSCRIPTIONS_DIALOG_UI_STATE,
+              true
+            );
+            return Promise.resolve();
+          }
+
+          // Switch to the paywall page instead.
+          return this.switchTo_(
+            this.pages_[PAYWALL_PAGE_INDEX].element.id,
+            direction
+          );
+        }
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+        // Navigate to the paywall page with blocked state, show the paywall after some delay.
+        this.paywallTimeout_ = setTimeout(() => {
+          this.storeService_.dispatch(
+            Action.TOGGLE_SUBSCRIPTIONS_DIALOG_UI_STATE,
+            true
+          );
+        }, PAYWALL_DELAY_DURATION);
+      } else {
```

Comment:
we should also reset page to navigate to here? Please add test for this

Reply:
e.g. user starts on page 7 but then clicks backwards to non-blocked content, back to paywall, logs in, shouldn't go to page 7

Reply:
Not sure what you mean. Once the status becomes granted, it would go back to the paywall page because the paywall page is assigned to `this.pageAfterGranted_ `.

Reply:
Missed the second comment but yes thanks for the catch! Fix this by resetting `this.pageAfterGranted_` when the ui state switches to false. Also added a unit test for this.